Draft-Accepted/RFCXXXX-Using-IDisposables.md

```diff
@@ -0,0 +1,399 @@
+try {
+    [Console]::WriteLine($reader.ReadLine())
+}
+finally {
+    $reader.Dispose()
+}
+```
+
+The drawbacks here are
+
++ The user needs to use a `try/finally` block to ensure it is cleaned up
++ The `finally` block can be interrupted with an engine stop request
++ The user need to explicitly write code not only to pre-define the vars, set the vars, then dispose of them with a null check
+
+C# developers can use the [using statement](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/using) that allows them to scope an `IDisposable` instance in an automatic `try/finally` block like the above:
+
+```csharp
+using (StreamReader reader = File.OpenText("numbers.txt"))
+{
+    Console.WriteLine(reader.ReadLine());
+} // reader is disposed here
+```
+
+This RFC proposes adding an equivalent `using` statement in PowerShell that can do the same thing as C#.
+For example the same C# `StreamReader` example in PowerShell can now be:
+
+```powershell
+using ($reader = [System.IO.File]::OpenRead("numbers.txt")) {
+    [Console]::WriteLine($reader.ReadLine())
+}
+```
+
+Another more complex example where the user might have multiple disposable objects that should be disposed is:
+
+```powershell
+$fs = $cryptoStream = $sr = $null
+try {
+    $fs = [System.IO.File]::OpenRead($Path)
+    $cryptoStream = [System.Security.CryptoGraphy.CryptoStream]::new(
+        $fs,
+        [System.Security.Cryptography.ToBase64Transform]::new(),
+        [System.Security.Cryptography.CryptoStreamMode]::Read)
+    $sr = [System.IO.StreamReader]::new($cryptoStream, [System.Text.Encoding]::ASCII)
+    $sr.ReadToEnd()
+}
+finally {
+    ${sr}?.Dispose()
+    ${cryptoStream}?.Dispose()
+    ${fs}?.Dispose()
+}
+```
+
+This is a more complicated version as now the finally block needs to check that each of the variables to be disposed are actually set and they are disposed in the opposite order they were defined.
+The proposed syntax for supporting this scenario would be:
+
+```powershell
+using (
+    $fs = [System.IO.File]::OpenRead($Path)
+    $cryptoStream = [System.Security.CryptoGraphy.CryptoStream]::new(
+        $fs,
+        [System.Security.Cryptography.ToBase64Transform]::new(),
+        [System.Security.Cryptography.CryptoStreamMode]::Read)
+    $sr = [System.IO.StreamReader]::new($cryptoStream, [System.Text.Encoding]::ASCII)
+) {
+    $sr.ReadToEnd()
+}
+```
+
+In this scenario the `using ()` statement supports multiple `AssignmentStatementAst` statements allowing it to track multiple variables to dispose.
+The equivalent syntax in C# would be this example:
+
+```csharp
+using (FileStream fs = File.OpenRead(path))
+using (CryptoStream cryptoStream = new CryptoStream(...))
+using (StreamReader sr = new StreamReader(cryptoStream, Encoding.ASCII))
+{
+    sr.ReadToEnd();
+}
+```
+
+## Specification
+
+The syntax for the proposed `Using` statement would be:
+
+```
+using ($<disposable>; ...) {
+    <Statement list>
+}
```

Comment:
perhaps a reasonable alternative

```powershell
using (<assignmentAst>)
statement-block
```

and allow a single assignment _only_, rather than a collection of assignments

Reply:
It's called out later on under [Multiple Disposables](https://github.com/PowerShell/PowerShell-RFC/pull/368/files/92b6a481b21f84dab7a08a490664b819c291074b#diff-6d3f6a46605e24b701b341f762f1d6cf1165758bc9dd7ad4650bd3a9ac602d39R222). The reason why I'm initially in favour of allowing a collection is if you are wrapping multiple disposable blocks you would essentially need to indent each using which is annoying to do.